src/include/RADOS/RADOS.hpp

```diff
@@ -0,0 +1,987 @@
+  WriteOp& operator =(WriteOp&&) = default;
+
+  void set_mtime(ceph::real_time t);
+  void create(bool exclusive);
+  void write(uint64_t off, ceph::buffer::list&& bl);
+  void write_full(ceph::buffer::list&& bl);
+  void writesame(std::uint64_t off, std::uint64_t write_len,
+		 ceph::buffer::list&& bl);
+  void append(ceph::buffer::list&& bl);
+  void remove();
+  void truncate(uint64_t off);
+  void zero(uint64_t off, uint64_t len);
+  void rmxattr(std::string_view name);
+  void setxattr(std::string_view name,
+		ceph::buffer::list&& bl);
+  void rollback(uint64_t snapid);
+  void set_omap(const boost::container::flat_map<std::string,
+		                                 ceph::buffer::list>& map);
+  void set_omap_header(ceph::buffer::list&& bl);
+  void clear_omap();
+  void rm_omap_keys(const boost::container::flat_set<std::string>& to_rm);
+  void set_alloc_hint(uint64_t expected_object_size,
+		      uint64_t expected_write_size,
+		      alloc_hint::alloc_hint_t flags);
+  void exec(std::string_view cls, std::string_view method,
+	    const ceph::buffer::list& inbl, boost::system::error_code* ec = nullptr);
+};
+
+// Come back and refactor this layer properly at some point.
+using EnumeratedObject = librados::ListObjectImpl;
+class EnumerationCursor final {
+public:
+  static EnumerationCursor begin();
+  static EnumerationCursor end();
+
+  EnumerationCursor();
+  EnumerationCursor(const EnumerationCursor&);
+  EnumerationCursor& operator =(const EnumerationCursor&);
+  EnumerationCursor(EnumerationCursor&&);
+  EnumerationCursor& operator =(EnumerationCursor&&);
+  ~EnumerationCursor();
+
+  friend bool operator ==(const EnumerationCursor& lhs,
+			  const EnumerationCursor& rhs);
+  friend bool operator !=(const EnumerationCursor& lhs,
+			  const EnumerationCursor& rhs);
+  friend bool operator <(const EnumerationCursor& lhs,
+			 const EnumerationCursor& rhs);
+  friend bool operator <=(const EnumerationCursor& lhs,
+			  const EnumerationCursor& rhs);
+  friend bool operator >=(const EnumerationCursor& lhs,
+			  const EnumerationCursor& rhs);
+  friend bool operator >(const EnumerationCursor& lhs,
+			 const EnumerationCursor& rhs);
+
+  std::string to_str() const;
+
+private:
+  struct end_magic_t {};
+  EnumerationCursor(end_magic_t);
+  EnumerationCursor(void*);
+  friend RADOS;
+  static constexpr std::size_t impl_size = 16 * 8;
+  std::aligned_storage_t<impl_size> impl;
+};
+
+class RADOS final
+{
+public:
+  static constexpr std::tuple<uint32_t, uint32_t, uint32_t> version() {
+    return {0, 0, 1};
+  }
+
+  class Builder {
+    std::optional<std::string> conf_files;
+    std::optional<std::string> cluster;
+    std::optional<std::string> name;
+    std::vector<std::pair<std::string, std::string>> configs;
+    bool no_default_conf = false;
+    bool no_mon_conf = false;
+
+  public:
+    Builder() = default;
+    void add_conf_file(std::string_view v);
+    void set_cluster(std::string_view c) {
```

Comment:
in the builder pattern, these setter functions commonly return `Builder&` so they can be chained together with the final call to `build()`

Reply:
Fixed!